core/roboconf-dm-monitoring/src/main/java/net/roboconf/dm/monitoring/internal/TemplateWatcher.java

```diff
@@ -0,0 +1,360 @@
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.filefilter.AbstractFileFilter;
+import org.apache.commons.io.filefilter.CanReadFileFilter;
+import org.apache.commons.io.filefilter.FileFilterUtils;
+import org.apache.commons.io.monitor.FileAlterationListenerAdaptor;
+import org.apache.commons.io.monitor.FileAlterationMonitor;
+import org.apache.commons.io.monitor.FileAlterationObserver;
+
+/**
+ * A file system watcher dedicated to the Roboconf application templates.
+ *
+ * @author Pierre Bourret - Université Joseph Fourier
+ */
+public class TemplateWatcher extends FileAlterationListenerAdaptor {
+
+	/**
+	 * The monitoring manager.
+	 */
+	private final MonitoringManager manager;
+
+	/**
+	 * The monitoring template directory.
+	 */
+	private final File templateDir;
+
+	/**
+	 * The template alteration monitor.
+	 */
+	private final FileAlterationMonitor monitor;
+
+	/**
+	 * Flag indicating if this watcher has already tracked down the templates already present in the template
+	 * directory.
+	 */
+	private AtomicBoolean hasBeenProvisioned = new AtomicBoolean( false );
+
+	/**
+	 * The lock for keeping watched templates.
+	 */
+	private final ReadWriteLock lock = new ReentrantReadWriteLock( true );
+
+	/**
+	 * The templates being watched, index by the name of the scoped application (or {@code null} for global templates)
+	 * and then by the template identifier.
+	 *
+	 * @GuardedBy lock
+	 */
+	private final Map<String, Map<String, TemplateEntry>> templates = new HashMap<String, Map<String, TemplateEntry>>();
+
+	/**
+	 * The Handlebars template processor.
+	 */
+	private final Handlebars handlebars = new Handlebars();
+
+	/**
+	 * The logger.
+	 */
+	private final Logger logger = Logger.getLogger( this.getClass().getName() );
+
+	/**
+	 * Create a template watcher.
+	 *
+	 * @param manager      the monitoring manager, to which event handling is delegated.
+	 * @param templateDir  the monitoring template directory to watch.
+	 * @param pollInterval the poll interval.
+	 * @throws IOException if there is a problem watching the template directory.
+	 */
+	public TemplateWatcher( final MonitoringManager manager, final File templateDir, final long pollInterval )
+			throws IOException {
+		this.manager = manager;
+
+		// Register the custom helpers.
+		this.handlebars.registerHelper( AllHelper.NAME, AllHelper.INSTANCE );
+
+		// We need the canonical directory, in order to test relationship w/ monitoring template candidates.
+		this.templateDir = templateDir.getCanonicalFile();
+
+		// Create the observer, register this object as the event listener.
+		final FileAlterationObserver observer = new FileAlterationObserver(
+				this.templateDir, TemplateEntry.getTemplateFileFilter( this.templateDir ) );
+		observer.addListener( this );
+
+		// Create the monitor.
+		this.monitor = new FileAlterationMonitor( pollInterval, observer );
+		this.monitor.setThreadFactory( new ThreadFactory() {
```

Comment:
It's even more/worse than that.
The thread factory is completely stateless, so we could use a constant one:
```java
private static final ThreadFactory THREAD_FACTORY = new ThreadFactory() {
    @Override
    public Thread newThread( final Runnable r ) {
        return new Thread( r, "Roboconf.TemplateWatcher" );
    }
});
```

But the problem remains: the use of anonymous evil classes...

So inner (named) class + constant? or constant = new anonymous ?
As some coding standards are still obscure to me, I let you choose. ;)